### ISSUE TITLE
push_rule API call causes 404 with selft-hosted gitlab CE

### ISSUE THREAD

**Issue by snake007v** [pulumi/pulumi-gitlab#514]
### What happened?

	In the case of self-hosted Gitlab CE it will cause a 404 error on push_rule API call as it's available only for EE.
	This call can't be turned off and the error can't be handled.

### Example

```
	// Create a new GitLab Group
	gitlabGroup, err := gitlab.NewGroup(ctx, gitlabGroupName, &gitlab.GroupArgs{
		Name: pulumi.String(gitlabGroupName),
		Path: pulumi.String(gitlabGroupName),
	}, pulumi.Provider(gitlabProvider))
	if err != nil {
		return nil, err
	}
```
```
Diagnostics:
  gitlab:index:Group (paas-vm2-test):
    error:   sdk/resource_gitlab_group.go:564: provider: [ERROR] Failed to get push rules for group: group_id=8081 error="GET https://git.xxx.com/api/v4/groups/8081/push_rule: 404 {error: 404 Not Found}" provider=gitlab@v6.8.1	
```

### Output of `pulumi about`

```
❯ pulumi about
CLI          
Version      3.103.1
Go Version   go1.21.5
Go Compiler  gc

Plugins
NAME  VERSION
go    unknown

Host     
OS       darwin
Version  14.1.1
Arch     arm64

This project is written in go: executable='/opt/homebrew/bin/go' version='go version go1.21.6 darwin/arm64'
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by iwahbe** [comment]
Hi @snake007v. Just to clarify, is the bug that a 404 is returned, or that the error can't be handled? 

> In the case of self-hosted Gitlab CE it will cause a 404 error on push_rule API call as it's available only for EE.

It sounds like the 404 is correct.

**Comment by snake007v** [comment]
Sorry if I wasn't clear in my initial report.
This bug is about the `push_rule` API call itself, which will be executed even if you don't configure any rules in your code(as you can see from my example). And there is no way to disable that call. 

As a result, we will always get 404 error in case if you're creating a new group in GitlabCE.

**Comment by iwahbe** [comment]
Thanks for clarifying.

**Comment by qerub** [comment]
For the record: I'm running into this issue on self-hosted GitLab *EE* too.

**Comment by iwahbe** [comment]
@Medialo thanks for the patch and @qerub thanks for letting us know it has merged. As soon as upstream releases it should be available in pulumi-gitlab.

**Comment by Zacharieg** [comment]
Hi everyone, it seems that the error still stick around, even in the last version of pulumi gitlab (which use 17.1.0 of terraform provider). Do you have any news on this bug ? Have a great day :grin:

**Comment by Medialo** [comment]
@Zacharieg I have no more error with my fix for my case with a gitlab CE for push rules. 
I am going to test again with last version of gitlab provider and try to check that !

**Comment by skycaptain** [comment]
Just encountered the same issue. Upgrading `pulumi-gitlab` from v7.0.2 to latest v8.1.0 fixed the issue for me.

**Comment by skycaptain** [comment]
I need to make a correction. It seems the error did not reoccur after re-running the Pulumi project. However, when deploying the stack from scratch, I still encounter even with the latest v8.1.0:

```plain
  gitlab:index:Group (group):
    error:   sdk/resource_gitlab_group.go:576: provider: [ERROR] Failed to get push rules for group: error="GET https://[redacted]/api/v4/groups/2944/push_rule: 404 {message: 404 Not Found}" provider=gitlab@v8.1.0 group_id=2944
```

**Comment by qerub** [comment]
@skycaptain: Is this with GitLab EE? The Terraform provider is implemented in such a way that it might log this error but continue executing: https://gitlab.com/gitlab-org/terraform-provider-gitlab/-/blob/0f680cfb7da857f233bee03e202f925345843ac6/internal/provider/sdk/resource_gitlab_group.go#L576-579

**Comment by skycaptain** [comment]
Yes, with v17.0.3-ee.

**Comment by Zacharieg** [comment]
Same for me, in GitLab EE 17.0... Is there a way to make those errors in warnings ? When creating several nested groups, it's really a problem because it does not create subgroups if parent creating failed (even if the parent is created)

**Comment by guineveresaenger** [comment]
Hi @skycaptain @Zacharieg - would it be possible for you to send us some logs for what you see?

- the verbose output of `pulumi up` - https://www.pulumi.com/docs/support/troubleshooting/#verbose-logging 
- and also the GRPC logs: `PULUMI_DEBUG_GRPC="grpc.jsonl"`

Just to clarify - this is resulting in a hard error for you on the Pulumi side but you'd like it logged as a warning, which is what Terraform does?

Thank you!

**Comment by Zacharieg** [comment]
In facts it appear that I was wrong, the execution does continue... So this is not a critical bug (but pretty annoying).

The error seems to be on GitLab side, when no push rules are available, it return a 404 code, it's weird. And in v17 of GitLab, groups push rules are no longer available and I see no mention of that in the [v17 changelog](https://docs.gitlab.com/ee/update/versions/gitlab_17_changes.html), except a [new role setting to manage push rule](https://about.gitlab.com/releases/2024/05/16/gitlab-17-0-released/#new-permissions-for-custom-roles) on ultimate version (and I'm not in this version)...

I think the correct behaviour that pulumi should have here is too ignore the 404 error on push_rule when getting group information, because it seems to be a normal return if the group has no push rule (which is the case of many groups).

Just for info, here is the log of pulumi up : 
[out.txt](https://github.com/user-attachments/files/16095906/out.txt)
[grpc.json](https://github.com/user-attachments/files/16096017/grpc.json)

**Comment by guineveresaenger** [comment]
Hi @Zacharieg - sorry for the late reply here.

Thank you for the logs and the confirmation that execution continues. 
Ideally, GitLab would of course not return an error here; which is the underlying bug.

This may be an instance of Pulumi too eagerly forwarding "errors" logged as warnings and we'll look into that as soon as we can.

**Comment by andiolsi** [comment]
Just fyi, the error disappeared for me after i just went and set `pushRules: undefined`. 
Or at least it didn't get logged, however you wanna put it. In any case: Annoyance -> gone.

**Comment by iwahbe** [comment]
We can't repro, so it's hard to confirm, but I'm pretty sure that https://github.com/pulumi/pulumi-terraform-bridge/pull/2686 will bring our behavior fully inline with the upstream provider here.

**Comment by lmilbaum** [comment]
The same error thrown with gitlab.com.  As a Pulumi newbie, I'm not sure if a solution is provided in this discussion.

**Comment by rafalkrupinski** [comment]
I got this error after upgrading the provider from 6.* to 8.8

Adding 
```js
pushRules: undefined,
```
to GroupArgs helped.

Thank you

**Comment by sockyone** [comment]
@rafalkrupinski it won't fix.
Still waiting for the solution from the team.

**Comment by skycaptain** [comment]
Just encountered the issue again with `provider=gitlab@8.11.0` when running `pulumi preview --refresh --expect-no-changes`

**Comment by antdking** [comment]
issue is still present with provider=gitlab@9.0.0, for groups in gitlab.com SaaS Ultimate plan

**Comment by qerub** [comment]
My MR was merged upstream (thanks @RicePatrick!) so the error log message should be gone once #1043 is merged.
